src/shared/logger.ts

```diff
@@ -0,0 +1,216 @@
+/*!
+ * Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+'use strict'
+
+import * as os from 'os'
+import * as path from 'path'
+import * as vscode from 'vscode'
+import * as nls from 'vscode-nls'
+import * as winston from 'winston'
+import * as Transport from 'winston-transport'
+import { extensionSettingsPrefix } from './constants'
+import { DefaultSettingsConfiguration, SettingsConfiguration } from './settingsConfiguration'
+
+const localize = nls.loadMessageBundle()
+
+const LOG_RELATIVE_PATH: string = path.join('Code', 'logs', 'aws_toolkit')
+const DEFAULT_LOG_LEVEL: string = 'info'
+const DEFAULT_LOG_NAME: string = `aws_toolkit_${makeDateString('filename')}.log`
+const DEFAULT_OUTPUT_CHANNEL: vscode.OutputChannel = vscode.window.createOutputChannel('AWS Toolkit Logs')
+
+let defaultLogger: Logger
+
+export interface Logger {
+    logPath?: string
+    outputChannel?: vscode.OutputChannel
+    level: string
+    debug(...message: ErrorOrString[]): void
+    verbose(...message: ErrorOrString[]): void
+    info(...message: ErrorOrString[]): void
+    warn(...message: ErrorOrString[]): void
+    error(...message: ErrorOrString[]): void
+    releaseLogger(): void
+}
+
+export interface LoggerParams {
+    outputChannel?: vscode.OutputChannel
+    logPath?: string
+    logLevel?: string
+}
+
+// creates default logger object
+// can overwrite defaultLogger if you include params
+export function initialize(params?: LoggerParams): Logger {
+    if (!params) {
+        defaultLogger = createLogger({
+            outputChannel: DEFAULT_OUTPUT_CHANNEL,
+            logPath: getDefaultLogPath()
+        })
+        // only the default logger (with default params) gets a registered command
+        vscode.commands.registerCommand(
```

Comment:
Trying to register a command that is already registered causes an error. It might be worth guarding against this if possible. [getCommands](https://code.visualstudio.com/api/references/vscode-api#commands) might help